### PR TITLE
update side input cache when context changes, fix #865

### LINKED ISSUE #865: Side input not refreshing
This is my code to refresh side inputs that fit in memory.  I implemented this in Java version of Apache Beam and it worked refreshing the side input values every 5 seconds.

```scala
 val sideInputValues
    = sc.customInput("GenerateSequence",
      GenerateSequence.from(0).withRate(1, Duration.standardSeconds(1)))
      .withFixedWindows(Duration.standardSeconds(5))
      .count
      .flatMap(e => {
        //get data from service
        val random = new Random()
        val az = random.nextInt(10000)
        val ca = random.nextInt(10000)
        logger.info("GenerateSequence: "
          + "AZ: " + az + " CA: " + ca)
        Seq("AZ" -> az, "CA" -> ca)
      }).asMapSideInput

    sc.customInput("Read From Pubsub", PubsubIO
      .readMessages()
      .withIdAttribute("id")
      .withTimestampAttribute("ts")
      .fromTopic("topic"))
      .withFixedWindows(Duration.standardSeconds(5))
      .flatMap(e =>
        Try(new String(e.getPayload)).toOption
      )
      .countByValue
      .withSideInputs(sideInputValues)
      .map((e, ctx) => {
        val sideInput = ctx(sideInputValues)
        logger.info("Track: " + e + " sideInput: "
          + "AZ: " + sideInput.get("AZ") + " CA: " + sideInput.get("CA"))
      })
    sc.close()
```

Java version
```java
 final PCollectionView<Map<String, Integer>> sideInput =
        pipeline.apply("GenerateSequence",
                       GenerateSequence.from(0).withRate(1, Duration.standardSeconds(1)))
            .apply("Window GenerateSequence",
                   Window.into(FixedWindows.of(Duration.standardSeconds(5))))
            .apply("Counts", Combine.globally(Sum.ofLongs()).withoutDefaults())
            .apply("RefreshCache", ParDo.of(new RefreshCache()))
            .apply("viewTags", View.asMap());

    final PubsubIO.Read<PubsubMessage> pubsubRead = PubsubIO
        .readMessages()
        .withIdAttribute("id")
        .withTimestampAttribute("ts")
        .fromTopic("topic");

    pipeline.apply("Read from", pubsubRead)
        .apply("Window Fixed",
               Window.into(FixedWindows.of(Duration.standardSeconds(5))))
        .apply("Get Metadata", ParDo.of(new GetMetadata()))
        .apply("Count Per Element", Count.perElement())
        .apply("Join", ParDo.of(new MetadataJoin(sideInput)).withSideInputs(sideInput));


private static class RefreshCache extends DoFn<Long, KV<String, Integer>> {

    private static final long serialVersionUID = 1;
    private static final Random RANDOM = new Random();

    @ProcessElement
    public void processElement(ProcessContext c) {
      KV<String, Integer> kv = KV.of("CA", RANDOM.nextInt(10000));
      c.output(kv);
    }
  }

 private static class MetadataJoin extends DoFn<KV<String, Long>, Void> {

    private static final long serialVersionUID = 1;
    private final PCollectionView<Map<String, Integer>> sideInput;
    private static final Logger LOG = LoggerFactory.getLogger(MetadataJoin.class);

    public MetadataJoin(PCollectionView<Map<String, Integer>> sideInput) {
      this.sideInput = sideInput;
    }

    @ProcessElement
    public void processElement(ProcessContext c) {
      final Map<String, Integer>
          sideInputValues =
          firstNonNull(c.sideInput(this.sideInput), new HashMap<>());
      final KV<String, Long> kv = c.element();

      final Integer ca = sideInputValues.get("CA");
      LOG.info("Metadata: {}, Side input value: {}", kv.getKey(), ca);
    }
  }
```

### PR DESCRIPTION
**Pull request by nevillelyh**

### CODE REVIEW

**Comment by ravwojdyla** [comment]
@nevillelyh when does this happen?

**Comment by nevillelyh** [comment]
See #865, there are 2 collections, both windowed. One is being used as a side input. Process context on the main collection changes everything a new window starts, and references the side input of the corresponding window.

File: scio-core/src/main/scala/com/spotify/scio/values/SideInput.scala [hunk 1]
```diff
@@ -30,12 +30,14 @@ import scala.collection.JavaConverters._
 trait SideInput[T] extends Serializable {
 
   private var cache: T = null.asInstanceOf[T]
+  private var context: AnyRef = null
```
Comment (ravwojdyla) [inline scio-core/src/main/scala/com/spotify/scio/values/SideInput.scala]: `null` to `_`  = `-3` characters :trollface:

**Comment by nevillelyh** [comment]
Gonna re-work this and submit a new PR later

**Pull request closed without merging**